src/gc-heap-snapshot.cpp

```diff
@@ -0,0 +1,495 @@
+
+struct HeapSnapshot {
+    vector<Node> nodes;
+    // edges are stored on each from_node
+
+    StringTable names;
+    StringTable node_types;
+    StringTable edge_types;
+    DenseMap<void*, size_t> node_ptr_to_index_map;
+
+    size_t num_edges = 0; // For metadata, updated as you add each edge. Needed because edges owned by nodes.
+};
+
+// global heap snapshot, mutated by garbage collector
+// when snapshotting is on.
+int gc_heap_snapshot_enabled = 0;
+HeapSnapshot *g_snapshot = nullptr;
+extern jl_mutex_t heapsnapshot_lock;
+
+void serialize_heap_snapshot(ios_t *stream, HeapSnapshot &snapshot, char all_one);
+static inline void _record_gc_edge(const char *node_type, const char *edge_type,
+                                   jl_value_t *a, jl_value_t *b, size_t name_or_index) JL_NOTSAFEPOINT;
+void _record_gc_just_edge(const char *edge_type, Node &from_node, size_t to_idx, size_t name_or_idx) JL_NOTSAFEPOINT;
+void _add_internal_root(HeapSnapshot *snapshot);
+
+
+JL_DLLEXPORT void jl_gc_take_heap_snapshot(ios_t *stream, char all_one) {
+    HeapSnapshot snapshot;
+    _add_internal_root(&snapshot);
+
+    jl_mutex_lock(&heapsnapshot_lock);
+
+    // Enable snapshotting
+    g_snapshot = &snapshot;
+    gc_heap_snapshot_enabled = true;
+
+    // Do a full GC mark (and incremental sweep), which will invoke our callbacks on `g_snapshot`
+    jl_gc_collect(JL_GC_FULL);
+
+    // Disable snapshotting
+    gc_heap_snapshot_enabled = false;
+    g_snapshot = nullptr;
+
+    jl_mutex_unlock(&heapsnapshot_lock);
+
+    // When we return, the snapshot is full
+    // Dump the snapshot
+    serialize_heap_snapshot((ios_t*)stream, snapshot, all_one);
+}
+
+// adds a node at id 0 which is the "uber root":
+// a synthetic node which points to all the GC roots.
+void _add_internal_root(HeapSnapshot *snapshot) {
+    Node internal_root{
+        snapshot->node_types.find_or_create_string_id("synthetic"),
+        snapshot->names.find_or_create_string_id(""), // name
+        0, // id
+        1, // size
+        0, // size_t trace_node_id (unused)
+        0, // int detachedness;  // 0 - unknown,  1 - attached;  2 - detached
+        vector<Edge>() // outgoing edges
+    };
+    snapshot->nodes.push_back(internal_root);
+}
+
+// mimicking https://github.com/nodejs/node/blob/5fd7a72e1c4fbaf37d3723c4c81dce35c149dc84/deps/v8/src/profiler/heap-snapshot-generator.cc#L597-L597
+// returns the index of the new node
+size_t record_node_to_gc_snapshot(jl_value_t *a) JL_NOTSAFEPOINT {
+    auto val = g_snapshot->node_ptr_to_index_map.insert(make_pair(a, g_snapshot->nodes.size()));
+    if (!val.second) {
+
+        return val.first->second;
+    }
+
+    ios_t str_;
+    bool ios_need_close = 0;
+
+    // Insert a new Node
+    size_t self_size = 0;
+    StringRef name = "<missing>";
+    StringRef node_type = "object";
+
+    jl_datatype_t* type = (jl_datatype_t*)jl_typeof(a);
+
+    if (type == (jl_datatype_t*)jl_buff_tag) {
+        name = "<buffer>";
+    }
```

Comment:
I think we removed all cases (in particular jl_buffer_type has this tag) that reached here?
```suggestion
```